File: calculator/BinarySort.java
public class BinarySort {

    /**
     * Sorts an array using Binary Insertion Sort.
     * Binary Insertion Sort improves upon Insertion Sort by using binary search 
     * to find the correct position to insert the current element.
     * 
     * @param array The array to be sorted.
     */
    public static void binaryInsertionSort(int[] array) {
        for (int i = 1; i < array.length; i++) {
            int key = array[i];
            int low = 0;
            int high = i - 1;

            // Binary search to find the position where key should be inserted
            while (low <= high) {
                int mid = low + (high - low) / 2;
                if (key < array[mid]) {
                    high = mid - 1;
                } else {
                    low = mid + 1;
                }
            }

            // Move all elements after the found position one position to the right
            for (int j = i - 1; j >= low; j--) {
                array[j + 1] = array[j];
            }
            array[low] = key;
        }
    }

    public static void main(String[] args) {
        int[] data = {37, 23, 0, 17, 12, 72, 31, 46, 100, 88, 54};
        System.out.println("Original array:");
        printArray(data);

        binaryInsertionSort(data);

        System.out.println("Sorted array:");
        printArray(data);
    }

    private static void printArray(int[] array) {
        for (int i : array) {
            System.out.print(i + " ");
        }
        System.out.println();
    }
}

File: calculator/binarysort.c
#include <stdio.h>

// Function to find the position where it should be inserted in sorted array
int binarySearch(int a[], int item, int low, int high) {
    while (low <= high) {
        int mid = low + (high - low) / 2;
        if (item == a[mid])
            return mid + 1;
        if (item > a[mid])
            low = mid + 1;
        else
            high = mid - 1;
    }
    return low;
}

// Function to sort array using Binary Insertion Sort
void binaryInsertionSort(int a[], int n) {
    int i, loc, j;

    for (i = 1; i < n; ++i) {
        loc = binarySearch(a, a[i], 0, i - 1);

        for (j = i - 1; j >= loc; --j)
            a[j + 1] = a[j];

        a[loc] = a[i];
    }
}

void printArray(int a[], int n) {
    for (int i = 0; i < n; ++i)
        printf("%d ", a[i]);
    printf("\n");
}

int main() {
    int a[] = {37, 23, 0, 17, 12, 72, 31, 46, 100, 88, 54};
    int n = sizeof(a) / sizeof(a[0]);

    printf("Original array: \n");
    printArray(a, n);

    binaryInsertionSort(a, n);

    printf("Sorted array: \n");
    printArray(a, n);

    return 0;
}
